qa/rpc-tests/test_framework/nodemessages.py

```diff
@@ -398,30 +398,59 @@ def ToHex(obj):
 
 # Objects that map to bitcoind objects, which can be serialized/deserialized
 
+# because the nVersion field has not been passed before the VERSION message the protocol uses an old format for the CAddress (missing nTime)
+# This class handles that old format
+class CAddressInVersion(object):
```

Comment:
The purpose of reducing code duplication is to minimize bugs caused by new changes -- in particular, people would make a change in one copy of the duplicated code but forget to do so in other copies.

In this case, that initial CAddress is frozen in time -- we can never allow it to change.  It would actually introduce a bug if changes to CAddress accidentally affected the format of the initial message.  

More generally, there is a strong argument that code supporting legacy or obsolete subsystems *should* be duplicated -- you basically need to be bug-for-bug compatible with the way things originally worked.  So isolating and freezing that code makes a lot of sense to ensure that when you add features to "living" code, you don't have to write compensation code that essentially un-does your new changes, and then prove that your change + compensation is bug-for-bug compatible with the legacy.

Reply:
:+1: